File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    let longest = 0;
    let currSubstr = new Set();
    
    for (let i = 0; i <= s.length; i++) {
        longest = Math.max(longest, currSubstr.size);
        
        if (currSubstr.has(s[i])) {
            const entries = currSubstr.entries();
            for (const [char, _] of entries) {
                currSubstr.delete(char);
                if (char === s[i]) break;
            }
        }
        
        currSubstr.add(s[i])
    }
    
    return longest;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        i = longest = 0
        charIdx = {}
        
        for j in range(len(s)):
            if s[j] in charIdx:
                i = max(charIdx[s[j]] + 1, i)
                
            longest = max(longest, j-i+1)
            charIdx[s[j]] = j
            
        return longest
        